Reports: Replace `Date Registered` column with `Last Activity`

m1r0 opened the pull request:
Part of #4833

### Changes proposed in this Pull Request

This PR replaces the "date registered" column in the students report with "last activity". The last activity date is based on lesson activities (lesson status change).

### Testing instructions

* Create two new courses with lessons.
* Go to the first course and mark a lesson as complete.
* Go to Sensei LMS -> Analysis.
* Make sure the last activity date and date diff for the first course is correct.
* Make sure the last activity is empty for the second course.
* Export the data to CSV and make sure it's correct.

m1r0 commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -465,6 +464,33 @@ protected function get_row_data( $item ) {
+				$last_activity = null;
+				if ( $last_activity_comment ) {
+					if ( $this->csv_output ) {
+						$last_activity = $last_activity_comment->comment_date;
```
Should we use GMT here or using WP timezone is good?

m1r0 replied:
The date registered was in GMT. Code updated in 510c9dddeae536273c6dfb322c1771ff68f5cb54.

donnapep commented:
> The last activity date is based on lesson activities (lesson status change).

I should have clarified this bit. Last activity is actually based on the date on which the last lesson was marked complete. So if I have two courses, and one lesson was completed in the first course yesterday, while another lesson was completed in the second course today, it would use the lesson completion date from the second course.

Also, could we get rid of the dotted underline here?

![Screen Shot 2022-02-23 at 10 51 02 AM](https://user-images.githubusercontent.com/1190420/155355458-465b8896-aed0-4511-bbae-d78c096dcb17.jpg)

donnapep commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -465,6 +464,33 @@ protected function get_row_data( $item ) {
+							sprintf(
+								/* translators: Time difference between two dates. %s: Number of seconds/minutes/etc. */
+								esc_html__( '%s ago', 'sensei-lms' ),
+								human_time_diff( strtotime( $last_activity_comment->comment_date_gmt ) )
```
Could we show the full date after 6 days of inactivity, and use `human_time_diff` otherwise?

donnapep commented:
One more thought. 😅  I think we should consider moving this functionality to its own function and add some tests for it.

m1r0 commented:
@donnapep I've addressed your suggestions in fd2002ca9b06580a623cf4432f43347872848671 and 2c6e83ac8b0c9d9c53a4bc2cadaf3d1f229c281a. Thanks for the help. 🙂

merkushin commented on tests/unit-tests/test-class-sensei-analysis-overview-list-table.php:
```diff
@@ -0,0 +1,149 @@
+
+		/* Act. */
+		// Simulate passing a quiz.
+		wp_update_comment(
```
Tests with `wp_update_comment` look like good candidates for a separate test with the data provider. 
All these tests look very similar with a difference in the in/out values.

m1r0 replied:
It's funny that you mention this because I usually split tests but here decided to not do it for some reason. 😄 
But I totally agree, splitting the tests makes them more readable and easier to maintain. Updated in 03caf181f0e9a94583aded134a7ee2d2221bbc06.

m1r0 replied:
I've done a few more test updates after the mentioned commit above.

donnapep reviewed (commented):
Left a few comments. Definitely going to borrow this code for my own PR. 🙂

donnapep commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -647,6 +646,52 @@ public function stats_boxes() {
+		);
+
+		if ( ! $last_activity_comment ) {
+			return '';
```
I believe we wanted to display "N/A" in this case.

m1r0 replied:
Updated in d0575dbd38c46d5cd3d01d9ee29e22bdca1756ff.

m1r0 commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -673,6 +646,52 @@ public function stats_boxes() {
+	 * @since 4.2.0
+	 *
+	 * @param  int $user_id
+	 * @return string
```
🤦 I will stop missing these one day, I promise! Fixed in d0575dbd38c46d5cd3d01d9ee29e22bdca1756ff.

m1r0 replied:
Just noticed that the "param comments check" is disabled in [phpcs](https://github.com/Automattic/sensei/blob/master/phpcs.xml.dist#L69). Maybe we should enable it? I'm not sure if there is a rule for `@return`.

donnapep replied:
> Just noticed that the "param comments check" is disabled in [phpcs](https://github.com/Automattic/sensei/blob/master/phpcs.xml.dist#L69). Maybe we should enable it?

Makes sense, so long as it would only impact changed code and not check the entire codebase.

m1r0 replied:
👍

 PR: https://github.com/Automattic/sensei/pull/4877

m1r0 commented on includes/class-sensei-analysis-overview-list-table.php:
```diff
@@ -673,6 +646,52 @@ public function stats_boxes() {
+				'number'  => 1,
+				'user_id' => $user_id,
+				'type'    => 'sensei_lesson_status',
+				'status'  => [ 'complete', 'passed', 'ungraded', 'failed' ],
```
Thanks for catching this. Fixed in 8049fdafd7d6f7a08d8753126d543f62b7820a0d and added more tests in 53a267928af1056397437be756005e53cec78632.

m1r0 commented on tests/unit-tests/test-class-sensei-analysis-overview-list-table.php:
```diff
@@ -0,0 +1,149 @@
+		);
+
+		/* Assert. */
+		$this->assertNotEmpty(
```
Good idea, updated in 88632499a52ae8adc9580c7b1278ac666c7fa1b8.

donnapep reviewed (approved):
If @merkushin is good with the tests, I'm good with everything else!
